### PR TITLE
[Inductor] Legalize BF16

### PR DESCRIPTION
**Pull request by EikanWang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96650
* #96620
* __->__ #96183
* #96485
* #96649

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -1777,7 +1866,38 @@ def to_dtype(x, dtype):
...
+                            V.graph.get_dtype(input_value.args[1])
+                            if input_value.target == "load"
+                            else input_value.args[-1]
+                        )
+                        if dtype in [torch.bfloat16]:
+                            opt_ctx.is_load_bf16_as_fp32 = True
+                        else:
+                            self.simd_vec = False
+                    elif dtype == torch.bfloat16:
+                        if not all(usr.target == "store" for usr in cur_node.users):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1885]: `reduction` is also a special store, consider to handle it here too, the logic can be shared.

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -1868,7 +1988,63 @@ def __init__(self, kernel_group):
...
+                    )
+
+                    if load_dtype == torch.bfloat16:
+                        with sub_graph.inserting_after(_node):
+                            to_type_node = sub_graph.call_method(
+                                "to_dtype", args=(ops, _node, torch.float)
+                            )
+                            to_type_node_args = to_type_node.args
+                            _node.replace_all_uses_with(to_type_node)
+                            to_type_node.args = to_type_node_args
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2012]: How do we handle the case where there is already `to_dtype` to `float` here. Will there be duplicated `to_dtype` on the graph and in the codegen?

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -1868,7 +1988,63 @@ def __init__(self, kernel_group):
         self.call_ranges = None
         self.picked_vec_isa: codecache.VecISA = codecache.pick_vec_isa()
 
+    def legalize_bf16(self, nodes):
+        def add_to_dtype(sub_graph: torch.fx.Graph):
+            for node in sub_graph.nodes:
+                _node: torch.fx.Node = node
+                if _node.target in ["load", "constant"]:
```
[the remark this answers is not in the document]
Comment (EikanWang) [reply torch/_inductor/codegen/cpp.py:1999]: The `dtype` of `index_expr` should be always interge. I suppose we don't need to handle it.

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (EikanWang) [reply torch/_inductor/codegen/cpp.py:2016]: Yes, there will duplicate `to_dtype` on the graph. I will submit another PR to eliminate the useless `to_dtype` nodes.

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py [hunk 4]
```diff
@@ -1868,7 +1994,96 @@ def __init__(self, kernel_group):
...
+                        ops,
+                        name,
+                        dtype,
+                        src_dtype,
+                        reduction_type,
+                        index,
+                        value,
+                    ) = _node.args
+                    if src_dtype == torch.bfloat16:
+                        # Since we always convert the load/store value to float if the tensor is blfoat16.
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2039]: nit: bfloat16

**Review by jgong5 — APPROVED** [review]
LGTM except for a nit.

File: torch/_inductor/codegen/cpp.py (hunk 4 above)
Comment (EikanWang) [reply torch/_inductor/codegen/cpp.py:2039]: Thanks. I will fix it in the next PR.

**Pull request closed without merging**